juju/juju
# [JUJU-2569] Add authentication checks for accessing cross model secrets

Access to cross model secrets is now gated by a macaroon. The macaroon used is that for the cross model relation to which the access grant for the secret is scoped. For a consumer to access a secret, these checks are performed:
- the consumer is in the relation to which the secret is scoped
- the relation is not suspended
- the macaroon grants access to the cross model relation

So secret access is denied if the consume permission is revoked or the relation is suspended.

A new cross model secret facade was needed to allow the consuming controller to get the scope to which the secret was shared. This then allows the correct macaroon to be selected for the subsequent call to get the content.

To allow for the offering controller to be temporarily unavailable, retry was added to the cross model api.

TODO
- add migration for cross secret consume records
- watch secret changes on offer side
- update consume counts on offer side 

## Checklist


## QA steps

deploy juju-qa-dummy-source and offer it
in a different model deploy juju-qa-dummy-sink and relate
create a secret in dummy-source and grant access to dummy-sink
in dummy-sink, get the secret by url, supplying a label
now get the secret again using just the label
on the offer model, suspend the relation
get the secret and ensure permission denied
